### PR TITLE
Added new command "sysinfo"under delloem commands

### PR DESCRIPTION
**Pull request by ravisaripalli**
Added new command "Sysinfo" under dell oem command. The purpose of the command is to get the system information.

### CODE REVIEW

**Comment by ravisaripalli** [comment]
Hi Alexander 

Can you please assign the reviewer for the this pull request.
Or Please do the code review.

Thanks
RaviSastry

**Comment by AlexanderAmelkin** [comment]
I'm actually already reviewing it. Have been busy with it since you originally posted the request. However, the amount of code is substantial, the code is quite dirty and doesn't conform to Coding Standards (even the original SourceForge ones, let alone the updated version), so it takes a long time to review and comment. Please be patient. The request is not abandoned and will be processed eventually.

File: lib/ipmi_delloem.c [hunk 1]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+	if (read_fru_area(intf, fru, id, i, fru_len, fru_data_tmp) < 0) {
+		free(fru_data);
+		return;
+	}
+	BoardOffsetPtr = (fru_data[BOARD_OFFSET])*OFFSETMULTIPLIER;
+	ProductOffsetPtr = (fru_data[PRODUCT_OFFSET])*OFFSETMULTIPLIER;
+	if(iDRAC_FLAG == IDRAC_14G)
+		language = fru_data[BoardOffsetPtr+BOARDLANGOFFSET];
+	else if (iDRAC_FLAG == IDRAC_13G)
+		language = fru_data[10];
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: So, iDRAC_13G's FRU is not IPMI-compliant? Or does it simply have Board Area starting at offset 8 ?
Anyway, most of this function's code looks like you're repeating in one way or another other FRU processings functions from `lib/ipmi_fru.c` and `lib/ipmi_ekanalyzer.c`. They already contain a lot of duplicating code and I wouldn't like to add more. Please think of reusing the existing code.

I especially dislike the manual calculation of area and field offsets for standardized areas. This definitely needs not to be reinvented.

File: lib/ipmi_delloem.c [hunk 2]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+	i++;	/* skip fru area version */
+	area_len = fru_data[i++] * 8; /* fru area length */
+	i++;	/* skip fru board language */
+	i += 3;	/* skip mfg. date time */
+
+	}
+
+	fru_area = get_fru_area_str(fru_data, &i);
+	if (fru_area != NULL && strlen(fru_area) > 0) {
+		free(fru_area);
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: Missing `fru_area = NULL` here.
The same applies to multiple cases below. Pointers MUST be set to NULL immediately after thy are `free()`'d.

File: lib/ipmi_delloem.c [hunk 3]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+	uint8_t data[4]={0};
+	struct FW_VER firmware_ver;
+
+	/* 20 80 01 52 51 9f a2 02 00 00 00 00 00 00 00 <- raw data*/
+	memset(&req, 0, sizeof(req));
+
+	if( (iDRAC_FLAG == IDRAC_12G) || (iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G) ) {
+		req.msg.netfn = DELL_OEM_NETFN;
+		req.msg.lun = 0;
+		req.msg.cmd = 0xbf;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: What's this magic `0xbf` ?

File: lib/ipmi_delloem.c [hunk 4]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+		req.msg.data = data;
+		req.msg.data_len = 1;
+		if((iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G))
+			data[0] = 0x04;
+		else
+			data[0] = 0x01;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: What are these magic numbers?

File: lib/ipmi_delloem.c [hunk 5]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+
+		rsp = intf->sendrecv(intf, &req);
+		if (rsp == NULL) {
+				printf(" Error Getting the Firmware version \n");
+				return -1;
+		}
+		if (rsp->ccode == 0xc1) {
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: This is `IPMI_CC_INV_CMD` from `ipmitool/ipmi_cc.h`.

File: lib/ipmi_delloem.c [hunk 6]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+			memset(&firmware_ver,0,sizeof(struct FW_VER));
+		}
+		else if (rsp->ccode > 0) {
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: And this is `IPMI_CC_OK`. However, better use just `if (rsp->ccode)`. This applies to other places in the code as well.

File: lib/ipmi_delloem.c [hunk 7]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+				printf(" Error getting device ID (%s)\n",
+					val2str(rsp->ccode, completion_code_vals));
+				return -1;
+		}
+		else
+			memcpy(&firmware_ver,(struct FW_VER*)rsp->data,sizeof(struct FW_VER));
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: There is no point in casting `rsp->data` to the structure pointer type. `memcpy()` expects `void *`.
Spaces are missing after commas.

File: lib/ipmi_delloem.c [hunk 8]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+
+	/* BCD value. byte->hex conversion used.*/
+	printf("IPMI Version        : %d.%d\n", dev_id->ipmi_version & 0x0F, dev_id->ipmi_version >> 4);
+	ipmi_version = dev_id->ipmi_version;
+
+
+	printf("Manufacturer ID     : %lu\n",
+		(uint32_t)(dev_id->manufacturer_id[2] & 0x0f) * 65536 +
+		(uint32_t)(dev_id->manufacturer_id[1] * 256) +
+		(uint32_t)dev_id->manufacturer_id[0]);
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: There is `IPMI_DEV_MANUFACTURER_ID()` for that. But yet again, I don't like the very idea of this function as it seems to copy the functionality of `ipmitool mc info`, `ipmitool fru print`, and/or `ipmitool ekanalyzer frushow`.

File: lib/ipmi_delloem.c [hunk 9]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+	}else if (rsp->data[0] == 0x57) {
+		printf("Status              : ");
+		if (rsp->data[1] & 0x01) {printf("Controller operational firmware corrupte "); }
+		if (rsp->data[1] & 0x02) {printf("Controller update 'boot block' firmware corrupte "); }
+		if (rsp->data[1] & 0x04) {printf("Internal Use Area of BMC FRU corrupted "); }
+		if (rsp->data[1] & 0x08) {printf("SDR Repository Empty "); }
+		if (rsp->data[1] & 0x10) {printf("IPMB signal lines do not respond "); }
+		if (rsp->data[1] & 0x20) {printf("Cannot access BMC FRU device "); }
+		if (rsp->data[1] & 0x40) {printf("Cannot access SDR Repository "); }
+		if (rsp->data[1] & 0x80) {printf("Cannot access SEL device"); }
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: No. Magic numbers are not acceptable. No `0x56`, `0x40` or whatever. Descriptive macros please.

File: lib/ipmi_delloem.c [hunk 10]
```diff
@@ -338,6 +365,689 @@ usage(void)
...
+	if (rsp->ccode > 0) {
+		printf(" Error getting device GUID (%s)\n",
+			val2str(rsp->ccode, completion_code_vals));
+		return -1;
+	}
+	printf("GUID                : %.2X%.2X%.2X%.2X-%.2X%.2X-%.2X%.2X-%.2X%.2X-%.2X%.2X%.2X%.2X%.2X%.2X\n",
+		rsp->data[15], rsp->data[14], rsp->data[13], rsp->data[12],
+		rsp->data[11], rsp->data[10], rsp->data[9], rsp->data[8],
+		rsp->data[7], rsp->data[6], rsp->data[5], rsp->data[4],
+		rsp->data[3], rsp->data[2], rsp->data[1], rsp->data[0]);
```
Comment (AlexanderAmelkin) [inline lib/ipmi_delloem.c]: Now this is copying `ipmitool mc guid`. Why?

Comment (AlexanderAmelkin) [reply lib/ipmi_delloem.c]: Especially taking in account #25

**Comment by AlexanderAmelkin** [comment]
You should NEVER merge upstream master into your pull request branch. Use `git rebase upstream/master` instead followed by `git push -f` to your branch.

None of the review comments have been addressed/answered so far. I will not merge this as it is. Sorry.

**Comment by ravisaripalli** [comment]
Hi Alexander 
After seeing code review comments on the proposed patch and to avoid the confusion ,
i am closing the pull request and will resubmit.

**Pull request closed without merging**